man/scan.man
.TH SCAN %manext1% 2020-02-17 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
scan \- produce a summary listing of nmh messages
.SH SYNOPSIS
.HP 5
.na
.B scan
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RI [ msgs ]
.RB [ \-clear " | " \-noclear ]
.RB [ \-form
.IR formatfile ]
.RB [ \-format
.IR string ]
.RB [ \-header " | " \-noheader ]
.RB [ \-width
.IR columns ]
.RB [ \-reverse " | " \-noreverse ]
.RB [ \-file
.IR filename ]
.ad
.SH DESCRIPTION
.B scan
produces a one-line-per-message listing of the specified
folder or messages.  Each
.B scan
line contains the message number
(name), the date, the \*(lqFrom:\*(rq field, the \*(lqSubject\*(rq field,
and, if room allows, some of the body of the message.  For example:
.PP
.RS 5
.nf
.ta \w'15+- 'u +\w'07/\|05x 'u +\w'Dcrocker  'u
15+	10/\|05 crocker	nned\0\0<<Last week I asked some of
16\-	10/\|05 crocker	message id format\0\0<<I recommend
18	10/\|06 brien	Re: Exit status from mkdir
19	10/\|07*brien	\*(lqscan\*(rq listing format in nmh
.fi
.RE
.PP
The `+' on message 15 indicates that it is the current message.
.PP
The `\-' on message 16 indicates that it has been replied to, as indicated
by a \*(lqReplied:\*(rq component (produced by the
.B \-annotate
switch
to the
.B repl
command).
.PP
The `*' on message 19 indicates that no \*(lqDate:\*(rq header was
present.  The time of last modification of the message is given instead.
.PP
If there is sufficient room left on the
.B scan
line after the
subject, the line will be filled with text from the body, preceded by
\*(lq<<\*(rq, and terminated by \*(lq>>\*(rq if the body is sufficiently short.
.B scan
actually reads each of the specified messages and parses them to extract
the desired fields.  During parsing, appropriate error messages will be
produced if there are format errors in any of the messages.
.PP
By default,
.B scan
will decode RFC 2047 (MIME) encoding in
these scan listings.
.B scan
will only decode these fields if your
terminal can natively display the character set used in the encoding.
You should set the appropriate
.IR locale (1)
environment variables to your native
character set, if it is not US-ASCII.  See
.IR locale (1)
for more details on the appropriate environment variables.
.PP
The switch
.BR \-reverse ,
makes
.B scan
list the messages in reverse
order.
.PP
The
.B \-file
.I filename
switch allows the user to obtain a
.B scan
listing of a mail drop.
This listing
includes every message in the mail drop (you can't scan individual messages).
The switch
.B \-reverse
is ignored with this option.
If
.I filename
is a file, it can be in
.B mbox
or
.B MMDF
format, as produced by
.BR packf .
If
.I filename
is a directory, it is considered to be in
.B Maildir
format.
.PP
The switch
.B \-width
.I columns
may be used to specify the width of
the scan line.  The default is to use the width of the terminal.
.PP
The
.B \-header
switch produces a header line prior to the
.B scan
listing.  Currently, the name of the folder and the current date and
time are output (see the
.B HISTORY
section for more information).
.PP
If the
.B \-clear
switch is used and
.BR scan 's
output is directed
to a terminal, then
.B scan
will consult the environment variables
.B $TERM
and
.B $TERMCAP
to determine your terminal type in order
to find out how to clear the screen prior to exiting.  If the
.B \-clear
switch is used and
.BR scan 's
output is not directed to a terminal
(e.g., a pipe or a file), then
.B scan
will send a formfeed prior
to exiting.
.PP
For example, the command:
.PP
.RS 5
(scan \-clear \-header; show all \-show pr \-f) | lpr
.RE
.PP
produces a scan listing of the current folder, followed by a formfeed,
followed by a formatted listing of all messages in the folder, one
per page.  Omitting
.RB \*(lq "\-show\ pr\ \-f" \*(rq
will cause the messages to be
concatenated, separated by a one-line header and two blank lines.
.PP
To override the output format used by
.BR scan ,
the
.B \-format
.I string
or
.B \-form
.I file
switches are used.  This permits individual fields of
the scan listing to be extracted with ease.  The string is simply a format
string and the file is simply a format file.  See
.IR mh\-format (5)
for the details.
.PP
In addition to the standard
.IR mh\-format (5)
escapes,
.B scan
also recognizes the following additional
.I component
escapes:
.PP
.RS 5
.nf
.ta \w'Dtimenow  'u +\w'Returns  'u
.I "Escape	Returns	Description"
body	string	the (compressed) first part of the body
dtimenow	date	the current date
folder	string	the name of the current folder
.fi
.RE
.PP
.RB { body }
returns a string without MIME decoding,
i.e.\& the MIME boundary separator may be the first text shown.
.PP
If no date header is present in the message, the
.I function
escapes
which operate on
.RB { date }
will return values for the date of last
modification of the message file itself.  This feature is handy for
scanning a draft folder, as message drafts usually aren't allowed
to have dates in them.
.PP
The
.B %nmhetcdir%
directory contains several format files as examples of customized
.B scan
output.
.PP
.B scan
will update the
.B nmh
context prior to starting the listing,
so interrupting a long
.B scan
listing preserves the new context.
.B nmh
purists hate this idea.
.SH FILES
.fc ^ ~
.nf
.ta \w'%nmhetcdir%/ExtraBigFileName  'u
^$HOME/.mh_profile~^The user profile
.fi
.SH "PROFILE COMPONENTS"
.fc ^ ~
.nf
.ta 2.4i
.ta \w'ExtraBigProfileName  'u
^Path:~^To determine the user's nmh directory
^Alternate\-Mailboxes:~^To determine the user's mailboxes
^Current\-Folder:~^To find the default current folder
.fi
.SH "SEE ALSO"
.IR pick (1),
.IR show (1),
.IR mh\-format (5)
.SH DEFAULTS
.nf
.RB ` +folder "' defaults to the current folder"
.RB ` msgs "' defaults to all"
.RB ` \-format "' defaulted as described above"
.RB ` \-noheader '
.RB ` \-width "' defaulted to the width of the terminal"
.fi
.SH CONTEXT
If a folder is given, it will become the current folder.
.SH HISTORY
Prior to using the format string mechanism,
.B \-header
used to generate
a heading saying what each column in the listing was.  Format strings
prevent this from happening.
.SH BUGS
The value of each
.I component
escape is set by
.B scan
to the
contents of the first message header
.B scan
encounters with the
corresponding component name; any following headers with the same
component name are ignored.


NEWS
#
# NEWS - Release notes for nmh 1.8
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been nearly five years since the last release of nmh, and there have
been a number of significant changes since that last release.  Long-time
MH and nmh users should read carefully the NOTABLE CHANGES section, as
there are some significant changes to nmh behavior.  Otherwise, please
see the README and INSTALL files for help on getting started with nmh.

This release is dedicated to Norman Z. Shapiro, co-designer of the MH
Message Handling System.  MH is the predecessor of nmh.  Norm was an
active supporter of nmh development until he passed away in October of
2021.  We are most grateful to Norm for his stewardship of MH and nmh.
    https://en.wikipedia.org/wiki/Norman_Shapiro

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

- Support for Content-MD5 header fields, MIME content cache functionality,
  and the message/partial MIME type have been removed.
- Gmail OAuth2/XOAUTH support for desktop applications has been effectively
  dropped, so nmh no longer supports it.  nmh support for Gmail API access
  is experimental, please post to nmh-workers@nongnu.org if you'd like to
  help with test and development.
- repl(1) -convertargs now allows editing of the composition draft between
  translation and any encoding of text content.  Because encoding can wrap
  long lines, the use of a paragraph formatter has been removed from
  mhn.defaults.

------------
NEW FEATURES
------------

- The default editor has been changed from 'vi' to 'prompter', to align with
  historical practice and reduce packaging dependencies on external programs.
- A new -checkbase64 switch has been added to mhfixmsg(1).
- inc(1)/msgchk(1) now support STARTTLS for the POP protocol.
- All TLS-supported protocols now will send the SNI (server name indicator)
  TLS extension.
- A new mh-format function %(ordinal) has been implemented to output the
  appropriate ordinal suffix for numbers.  Example: "%(num 22)%(ordinal)"
  will output "22nd".
- show and mhl now decode more addresses in header fields.
- Added warning from all programs that read the profile if the profile
  contains a post entry, which is ignored, but does not contain a
  postproc entry.  In other words, if you get this warning and want
  to suppress it, your options include:
  1) Remove your post profile entry.
  2) Make your post profile entry a comment by prepending it with the #:
     comment indicator.
  3) Add a postproc entry that points to the post that you use.  That can
     be viewed with "mhparam postproc".
- scan(1) -file argument can be a Maildir directory.
- Updated mhn.defaults to prefer mpv(1) over xv(1) and replace mpeg_play(1),
  and to use it for all video types, not just video/mpeg.  And prefer all
  other searched-for pdf viewers over acroread(1).
- Added mhshow-suffix-video.mp4 to mhn.defaults, for use by mhshow(1) and
  send(1).
- Removed support from mhn.defaults for application/x-ivs and text/richtext.
- Changed interpretation of argument to mhfixmsg(1) -decodeheaderfieldbodies
  switch to specify character set of the decoded field bodies.
- repl(1) -convertargs now allows editing of the composition draft between
  translation and any encoding of text content.
- install-mh(1) now enables the mh-draft(5) draft folder facility.

-----------------
OBSOLETE FEATURES
-----------------

- The generation and verification of a Content-MD5 field has been removed
  without deprecation.  The related -check and -nocheck options now error.
- The MIME content cache functionality has been mostly non-functional since
  the development on nmh, and consequently all of the content caching code
  and related switches (-cache/-rcache/-wcache) have been removed.
- Support for generating and reassembling message/partial messages has been
  removed; it seems that this has been broken since 1.5 and there is very
  little support across MUAs.
- Marked Gmail OAuth2/XOAUTH support as being unsupported.
- Support for the MHPDEBUG environment variable was removed.  It was
  deprecated in nmh 1.7.  The pick(1) -debug switch replaced it.
- The 'libdir' mhparam(1) component was removed.  It was deprecated in
  nmh 1.7, when it was replaced by a new 'libexecdir' component.

---------
BUG FIXES
---------

- Fixed bcc to work with sendmail/pipe, and better documented that dcc
  doesn't work with it [Bug 55700].
- An -attendee switch has been added to mhical(1), for use when more than one
  (or zero) attendees match a user's mailbox.
- Fixed inc(1) and %(me) function escape to not obey Local-Mailbox profile
  component.
- Fixed source charset in mhfixmsg textcharset verbose output.
- Fixed mhfixmsg charset determination of content added with -reformat.
- Fixed file descriptor leak in mhfixmsg when run on multiple input files.
- Fixed mhfixmsg(1) -decodeheaderfilebodies to support mixed encoded/undecoded.
- Fixed memory corruption in post(1) and inc(1) when using XOAUTH2,
  with 4 or more entries in the oauth-authservice file.
- Added alias expansion to From: address for use by sendfrom.
- Removed extra space added before header field bodies by dist(1) to $mhdraft.
- Fixed display of iCalendar object with multiple VEVENTS using a VTIMEZONE.
- Fixed allowable encodings with MIME message types to get closer to RFC 2046.
- Detect other files regardless of backup prefix [Bug #49476].
- Copy if hard link by refile(1) or send(1) fails with EACESS [Bug #56575].

----------------------------------------------------------------------------

#
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time

uip/scan.c
/* scan.c -- display a one-line "scan" listing of folder or messages
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include "h/mh.h"
#include "sbr/charstring.h"
#include "sbr/fmt_new.h"
#include "sbr/dtime.h"
#include "scansbr.h"
#include "sbr/m_name.h"
#include "sbr/getarguments.h"
#include "sbr/seq_setprev.h"
#include "sbr/seq_save.h"
#include "sbr/smatch.h"
#include "sbr/m_convert.h"
#include "sbr/getfolder.h"
#include "sbr/folder_read.h"
#include "sbr/folder_free.h"
#include "sbr/context_save.h"
#include "sbr/context_replace.h"
#include "sbr/context_find.h"
#include "sbr/brkstring.h"
#include "sbr/ambigsw.h"
#include "sbr/path.h"
#include "sbr/print_version.h"
#include "sbr/print_help.h"
#include "sbr/seq_getnum.h"
#include "sbr/error.h"
#include "sbr/fmt_compile.h"
#include "sbr/fmt_scan.h"
#include "h/tws.h"
#include "sbr/mts.h"
#include "sbr/done.h"
#include "sbr/utils.h"
#include "sbr/m_maildir.h"
#include "sbr/terminal.h"
#include "sbr/maildir_read_and_sort.h"
#include "sbr/globals.h"

#define SCAN_SWITCHES \
    X("clear", 0, CLRSW) \
    X("noclear", 0, NCLRSW) \
    X("form formatfile", 0, FORMSW) \
    X("format string", 5, FMTSW) \
    X("header", 0, HEADSW) \
    X("noheader", 0, NHEADSW) \
    X("width columns", 0, WIDTHSW) \
    X("reverse", 0, REVSW) \
    X("noreverse", 0, NREVSW) \
    X("file file", 4, FILESW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(SCAN);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(SCAN, switches);
#undef X


int
main (int argc, char **argv)
{
    bool clearflag = false;
    bool hdrflag = false;
    int ontty = 0;
    int width = -1;
    bool revflag = false;
    int i, state, msgnum;
    ivector_t seqnum = ivector_create (0);
    bool unseen;
    int num_unseen_seq = 0;
    char *cp, *maildir, *file = NULL, *folder = NULL;
    char *form = NULL, *format = NULL, buf[BUFSIZ];
    char **argp, *nfs, **arguments;
    struct msgs_array msgs = { 0, 0, NULL };
    struct msgs *mp;
    charstring_t scanl = NULL;
    FILE *in;

    if (nmh_init(argv[0], true, true)) { return 1; }

    mts_init ();
    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    /*
     * Parse arguments
     */
    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW:
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW:
		    die("-%s unknown", cp);

		case HELPSW:
		    snprintf (buf, sizeof(buf), "%s [+folder] [msgs] [switches]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case CLRSW:
		    clearflag = true;
		    continue;
		case NCLRSW:
		    clearflag = false;
		    continue;

		case FORMSW:
		    if (!(form = *argp++) || *form == '-')
			die("missing argument to %s", argp[-2]);
		    format = NULL;
		    continue;
		case FMTSW:
		    if (!(format = *argp++) || *format == '-')
			die("missing argument to %s", argp[-2]);
		    form = NULL;
		    continue;

		case HEADSW:
		    hdrflag = true;
		    continue;
		case NHEADSW:
		    hdrflag = false;
		    continue;

		case WIDTHSW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    width = atoi (cp);
		    continue;
		case REVSW:
		    revflag = true;
		    continue;
		case NREVSW:
		    revflag = false;
		    continue;

		case FILESW:
		    if (!(cp = *argp++) || (cp[0] == '-' && cp[1]))
			die("missing argument to %s", argp[-2]);
		    if (strcmp (file = cp, "-"))
			file = path (cp, TFILE);
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		die("only one folder at a time!");
            folder = pluspath (cp);
	} else
            app_msgarg(&msgs, cp);
    }

    if (!context_find ("path"))
	free (path ("./", TFOLDER));

    /*
     * Get new format string.  Must be before chdir().
     */
    nfs = new_fs (form, format, FORMAT);

    if (file) {
        /*
         * We have a -file argument
         */
	if (msgs.size)
	    die("\"msgs\" not allowed with -file");
	if (folder)
	    die("\"+folder\" not allowed with -file");

	/* check if "file" is really stdin */
	if (strcmp (file, "-") == 0) {
	    in = stdin;
	    file = "stdin";
	} else {
	    /* check if "file" is really a Maildir folder */
	    struct stat st;
	    if (stat (file, &st) == -1)
		adios (file, "unable to stat");
	    if (!(st.st_mode & S_IFDIR)) {
		if ((in = fopen (file, "r")) == NULL)
		    adios (file, "unable to open");
	    } else {
		/*
		 * We are scanning a Maildir folder
		 */
		struct Maildir_entry *Maildir;
		int num_maildir_entries;
		int msgnum = 0;
		char *fnp;
		FILE *in;
		int i;

		maildir_read_and_sort(file, &Maildir, &num_maildir_entries);
		for (i = 0; i < num_maildir_entries; i++) {
		    msgnum++;
		    fnp = Maildir[i].filename;

		    if ((in = fopen (fnp, "r")) == NULL) {
			admonish (fnp, "unable to open message");
			continue;
		    }

		    switch (state = scan (in, msgnum, 0, nfs, width,
					  0, 0, hdrflag ? file : NULL,
					  0L, 1, &scanl)) {
		    case SCNMSG:
		    case SCNERR:
			break;

		    default:
			die("scan() botch (%d)", state);

		    case SCNEOF:
			inform("message %d: empty", msgnum);
			break;
		    }
		    if (scanl)
			charstring_clear(scanl);
		    scan_finished ();
		    hdrflag = false;
		    fclose (in);
		    if (ontty)
			fflush (stdout);
		}
		done (0);
	    }
	}

	/*
	 * We are scanning a maildrop file
	 */
	if (hdrflag) {
	    printf ("FOLDER %s\t%s\n", file, dtimenow (1));
	}

	scan_detect_mbox_style (in);
	for (msgnum = 1; ; ++msgnum) {
	    state = scan (in, msgnum, -1, nfs, width, 0, 0,
			  hdrflag ? file : NULL, 0L, 1, &scanl);
	    if (scanl)
		charstring_clear(scanl);
	    if (state != SCNMSG)
		break;
	}
	charstring_free (scanl);
	scan_finished ();
	fclose (in);
	done (0);
    }

    /*
     * We are scanning a folder
     */

    if (!msgs.size)
	app_msgarg(&msgs, "all");
    if (!folder)
	folder = getfolder (1);
    maildir = m_maildir (folder);

    if (chdir (maildir) == -1)
	adios (maildir, "unable to change directory to");

    /* read folder and create message structure */
    if (!(mp = folder_read (folder, 1)))
	die("unable to read folder %s", folder);

    /* check for empty folder */
    if (mp->nummsg == 0)
	die("no messages in %s", folder);

    /* parse all the message ranges/sequences and set SELECTED */
    for (msgnum = 0; msgnum < msgs.size; msgnum++)
	if (!m_convert (mp, msgs.msgs[msgnum]))
	    done(1);
    seq_setprev (mp);			/* set the Previous-Sequence */

    context_replace (pfolder, folder);	/* update current folder         */
    seq_save (mp);			/* synchronize message sequences */
    context_save ();			/* save the context file         */

    /*
     * Get the sequence number for each sequence
     * specified by Unseen-Sequence
     */
    if ((cp = context_find (usequence)) && *cp) {
	char **ap, *dp;

	dp = mh_xstrdup(cp);
	ap = brkstring (dp, " ", "\n");
	for (i = 0; ap && *ap; i++, ap++)
	    ivector_push_back (seqnum, seq_getnum (mp, *ap));

	num_unseen_seq = i;
        free(dp);
    }

    ontty = isatty (fileno (stdout));

    for (msgnum = revflag ? mp->hghsel : mp->lowsel;
	 (revflag ? msgnum >= mp->lowsel : msgnum <= mp->hghsel);
	 msgnum += (revflag ? -1 : 1)) {
	if (is_selected(mp, msgnum)) {

	    if ((in = fopen (cp = m_name (msgnum), "r")) == NULL) {
		    admonish (cp, "unable to open message");
		continue;
	    }

	    if (hdrflag) {
		printf ("FOLDER %s\t%s\n", folder, dtimenow(1));
	    }

	    /*
	     * Check if message is in any sequence given
	     * by Unseen-Sequence profile entry.
	     */
	    unseen = false;
	    for (i = 0; i < num_unseen_seq; i++) {
		if (in_sequence(mp, ivector_at (seqnum, i), msgnum)) {
		    unseen = true;
		    break;
		}
	    }

	    switch (state = scan (in, msgnum, 0, nfs, width,
			msgnum == mp->curmsg, unseen,
			folder, 0L, 1, &scanl)) {
		case SCNMSG:
		case SCNERR:
		    break;

		default:
		    die("scan() botch (%d)", state);

		case SCNEOF:
		    inform("message %d: empty", msgnum);
		    break;
	    }
	    if (scanl)
		charstring_clear(scanl);
	    scan_finished ();
	    hdrflag = false;
	    fclose (in);
	    if (ontty)
		fflush (stdout);
	}
    }
    charstring_free (scanl);

    ivector_free (seqnum);
    folder_free (mp);	/* free folder/message structure */
    if (clearflag)
	nmh_clear_screen ();

    done (0);
    return 1;
}
